File: backend/app/imports.py
from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from sqlalchemy.orm import Session

from app.auth import get_current_account
from app.deps import get_session
from app.models import Account
from app.schemas import (
    ImportBatchRevalidationRequest,
    ImportConfirmRequest,
    ImportPreview,
    ImportRowRevalidation,
    ImportRowValidation,
    ImportRowValidationRequest,
    TransactionOut,
)
from app.services import imports as import_service
from app.transactions import _transaction_out, _write_warning

router = APIRouter(prefix="/import", tags=["import"])


@router.post("/preview", response_model=ImportPreview)
def preview_import(
    file: UploadFile = File(...),
    account: Account = Depends(get_current_account),
    session: Session = Depends(get_session),
) -> ImportPreview:
    """Parse and validate an uploaded .csv or .xlsx against the fixed template
    (T13): the response is the preview the user confirms — every row with its
    verdict (ok / error / duplicate). Nothing is inserted by this step."""
    content = file.file.read()
    if not content:
        raise HTTPException(status_code=422, detail="The file is empty")
    try:
        raw_rows = import_service.parse_file(file.filename or "", content)
    except (import_service.UnsupportedFile, import_service.BadTemplate) as error:
        raise HTTPException(status_code=422, detail=str(error))
    rows = import_service.preview_rows(session, account.id, raw_rows)
    return ImportPreview(
        rows=rows,
        ok_count=sum(1 for row in rows if row.status == "ok"),
        error_count=sum(1 for row in rows if row.status == "error"),
        duplicate_count=sum(1 for row in rows if row.status == "duplicate"),
    )


@router.post("/validate-row", response_model=ImportRowValidation)
def validate_row(
    payload: ImportRowValidationRequest,
    account: Account = Depends(get_current_account),
    session: Session = Depends(get_session),
) -> ImportRowValidation:
    """Re-validate one edited row during Verification (issue #44): the row's
    Wallet/Category names are re-resolved against the Account, the CONTEXT.md
    rules re-run, and the Duplicate check applied with the final key — the
    row's fresh status and message, through the same pipeline as the Preview.
    Nothing is written."""
    return import_service.revalidate_row(
        session, account.id, payload.row, payload.earlier_rows
    )


@router.post("/revalidate-rows", response_model=list[ImportRowRevalidation])
def revalidate_rows(
    payload: ImportBatchRevalidationRequest,
    account: Account = Depends(get_current_account),
    session: Session = Depends(get_session),
) -> list[ImportRowRevalidation]:
    """Batch Revalidation (issue #76): the draft's rows plus the target row
    numbers in, every target's fresh verdict out — one call through the same
    resolution, rules, and Duplicate key as the Preview and the single-row
    re-validation, including the in-file Duplicate check against preceding
    rows. Nothing is written."""
    known = {item.row for item in payload.rows if item.row is not None}
    unknown = [str(target) for target in payload.targets if target not in known]
    if unknown:
        raise HTTPException(
            status_code=422,
            detail=f"Unknown target row(s): {', '.join(unknown)}",
        )
    return import_service.revalidate_rows(
        session, account.id, payload.rows, payload.targets
    )


@router.post("/confirm", response_model=list[TransactionOut], status_code=201)
def confirm_import(
    payload: ImportConfirmRequest,
    account: Account = Depends(get_current_account),
    session: Session = Depends(get_session),
) -> list[TransactionOut]:
    """Insert the confirmed rows in one transaction (T13). Any invalid or
    now-duplicate row rejects the whole batch with 422: nothing is written
    until the user confirms."""
    try:
        created = import_service.confirm_rows(session, account.id, payload.rows)
    except import_service.ImportValidationError as error:
        raise HTTPException(status_code=422, detail=str(error))
    return [
        _transaction_out(
            session,
            account,
            transaction,
            warning=_write_warning(session, account, transaction),
        )
        for transaction in created
    ]

File: frontend/src/api/imports.ts
/** Import resource: upload / validate / confirm against the fixed template
 * (T13, issue #17). Import errors surface the backend's detail (e.g. "Unknown
 * wallet 'X'"), which is why these calls set `readDetail`. */

import { request } from './transport'
import type { Transaction } from './transactions'

export type ImportRowStatus = 'ok' | 'error' | 'duplicate'

export type ImportRow = {
  row: number
  status: ImportRowStatus
  type: string | null
  date: string | null
  amount: string | null
  wallet: string | null
  source_wallet: string | null
  destination_wallet: string | null
  category: string | null
  description: string | null
  latitude: string | null
  longitude: string | null
  error: string | null
}

export type ImportPreview = {
  rows: ImportRow[]
  ok_count: number
  error_count: number
  duplicate_count: number
}

/** A row the user kept, echoed back for confirmation: names, not ids — the
 * backend re-resolves them and re-runs every rule before writing anything. */
export type ImportRowInput = {
  row?: number
  type: 'expense' | 'income' | 'transfer'
  date: string
  amount: string
  wallet: string | null
  source_wallet: string | null
  destination_wallet: string | null
  category: string | null
  description: string | null
  latitude: string | null
  longitude: string | null
}

export async function previewImport(token: string, file: File): Promise<ImportPreview> {
  const form = new FormData()
  form.append('file', file, file.name)
  const response = await request('/import/preview', {
    method: 'POST',
    token,
    formData: form,
    readDetail: true,
    errorMessage: 'Could not read the file',
  })
  return (await response.json()) as ImportPreview
}

/** The fresh verdict for one edited row (issue #44): `status` speaks the
 * Preview's vocabulary — ok (Ready in the UI), duplicate, or error — and
 * `error` carries the message for an error row. */
export type ImportRowValidation = {
  status: ImportRowStatus
  error: string | null
}

/** Re-validate one edited row during Verification (issue #44/#46): the
 * row's Wallet/Category names are re-resolved against the Account, every
 * CONTEXT.md rule re-run, and the Duplicate check applied with the final
 * key. `earlierRows` is the draft's rows that precede it in the file — the
 * in-file Duplicate context — with their edits applied. Nothing is
 * written. */
export async function validateImportRow(
  token: string,
  row: ImportRowInput,
  earlierRows: ImportRowInput[],
): Promise<ImportRowValidation> {
  const response = await request('/import/validate-row', {
    method: 'POST',
    token,
    json: { row, earlier_rows: earlierRows },
    readDetail: true,
    errorMessage: 'Could not validate the row',
  })
  return (await response.json()) as ImportRowValidation
}

/** The fresh verdict for one target row of a batch Revalidation (issue
 * #76): `row` echoes the target's row number so the client can map each
 * verdict back to its draft row; `status` and `error` speak the Preview's
 * vocabulary, exactly like the single-row re-validation. */
export type ImportRowRevalidation = {
  row: number
  status: ImportRowStatus
  error: string | null
}

/** Batch Revalidation (issue #76): the draft's rows (with the user's edits
 * applied) plus the target row numbers in, every target's fresh verdict out
 * — one call through the same pipeline as the single-row re-validation,
 * including the in-file Duplicate context from preceding rows. Nothing is
 * written. */
export async function revalidateImportRows(
  token: string,
  rows: ImportRowInput[],
  targets: number[],
): Promise<ImportRowRevalidation[]> {
  const response = await request('/import/revalidate-rows', {
    method: 'POST',
    token,
    json: { rows, targets },
    readDetail: true,
    errorMessage: 'Could not re-validate the rows',
  })
  return (await response.json()) as ImportRowRevalidation[]
}

export async function confirmImport(
  token: string,
  rows: ImportRowInput[],
): Promise<Transaction[]> {
  const response = await request('/import/confirm', {
    method: 'POST',
    token,
    json: { rows },
    readDetail: true,
    errorMessage: 'Could not confirm the import',
  })
  return (await response.json()) as Transaction[]
}
